Training/box_it.cpp

```diff
@@ -1,100 +1,108 @@
+	Box (int length, int breadth, int height) : m_l(length), m_b(breadth), m_h(height) {
+	}
+
+	/// Es-Note: You can re-use constructors like this:
+	Box (const Box & b) : Box(b.m_l, b.m_b, b.m_h) {
+	}
+
+	int getLength ()  const  { return m_l; }
+	int getBreadth () const  { return m_b; }
+	int getHeight ()  const  { return m_h; }
+
+	long long CalculateVolume() const { return ((long long)m_l * m_b) * m_h; }
 };
 
 // Overload the operator < for the class Box. Box A < Box B if:
 // A.l < B.l
 // A.b < B.b and A.l == B.l
 // A.h < B.h and A.b == B.b and A.l == B.l
-bool operator<(Box& b1, Box& b2)
+bool operator< (const Box & a, const Box & b)
 {
-    int b1l = b1.getLength(), b2l = b2.getLength(),
-        b1b = b1.getBreadth(), b2b = b2.getBreadth(),
-        b1h = b1.getHeight(), b2h = b2.getHeight();
-    return b1l < b2l || (b1l == b2l && (b1b < b2b || (b1b == b2b && b1h < b2h)));
+	///Es-Notes: Const should help the compiler realize that these
+	///variables can be optimized away. You will see them in the disassembly
+	///when in debug mode, but in release they shouldn't be there.
+	const int al = a.getLength(),  bl = b.getLength(),
+			  ab = a.getBreadth(), bb = b.getBreadth(),
+			  ah = a.getHeight(),  bh = b.getHeight();
+
+	///Es-Notes: More short-circuiting opportunities w/more OR statements.
+	///Compiler will see that al == bl has already been determined.
+	return al  < bl
+		|| al == bl && ab  < bb
+		|| al == bl && ab == bb && ah < bh;
 }
 
 // Overload operator << for the class Box().
 // If B is an object of class Box :
 // cout << B should print B.l, B.b, and B.h on a single line separated by
 // spaces.
-ostream& operator<<(ostream& os, Box& b)
+ostream & operator<< (ostream & os, Box & b)
 {
-    os << b.getLength() << " " << b.getBreadth() << " " << b.getHeight();
-    return os;
+	os << b.getLength() << " " << b.getBreadth() << " " << b.getHeight();
+	return os;
 }
 
-int hackerrank::box_it() {
-    // START HACKERRANK BOILER PLATE
-    int n;
-    cin >> n;
-    Box temp;
-    for (int i = 0;i<n;i++)
-    {
-        int type;
-        cin >> type;
-        if (type == 1)
-        {
-            cout << temp << endl;
-        }
-        if (type == 2)
-        {
-            int l, b, h;
-            cin >> l >> b >> h;
-            Box NewBox(l, b, h);
-            temp = NewBox;
-            cout << temp << endl;
-        }
-        if (type == 3)
-        {
-            int l, b, h;
-            cin >> l >> b >> h;
-            Box NewBox(l, b, h);
-            if (NewBox<temp)
-            {
-                cout << "Lesser\n";
-            }
-            else
-            {
-                cout << "Greater\n";
-            }
-        }
-        if (type == 4)
-        {
-            cout << temp.CalculateVolume() << endl;
-        }
-        if (type == 5)
-        {
-            Box NewBox(temp);
-            cout << NewBox << endl;
-        }
-    }
-    // END HACKERRANK BOILER PLATE
-    return 0;
+int main () {
```

Comment:
Oops.

Reply:
Yeah, np. I'm basically transcribing these recommendations from the PR anyway. I figured that you wanted to actually run this one.

Reply:
On my linux VM I have the `main` functions in the files because there isn't a "solution" concept. I'm just one-off compiling the files as I finish them.